### PR TITLE
Add View#cached

### PR DESCRIPTION
**Pull request by NthPortal**
Resolves scala/bug#11168

### CODE REVIEW

**Comment by NthPortal** [comment]
I went with `cached` because it doesn't modify the view

**Review by diesalbla — APPROVED** [review]

**Comment by diesalbla** [comment]
For some unit testing, could we add a test like the one in [the bug or feature request](https://github.com/scala/bug/issues/11168#issue-362867355), only that instead of printing it would  increase a mutable counter?

**Comment by joshlemer** [comment]
Maybe in View.Id something like :

```scala
override def cached: View[A] = underlying match {
  case _: List[A] | _: Vector[A] | _: AnyOtherStrictlyEvaluatedCollection[A] => this
  case _ => super.cached
}
```

**Comment by joshlemer** [comment]
Would also need to make overrides for MapView and IndexedSeqView, here's my stab at it (feel free to change it if you don't like it)

```scala
  // in object IndexedSeqView
  class Cached[+A](underlying: SomeIndexedSeqOps[A]) extends IndexedSeqView[A] {

    private[this] lazy val iter = Iterator.empty ++ underlying.iterator

    private[this] lazy val array = new Array[Any](underlying.length)
    private[this] var nextIndex = 0

    /** Forces evaluation up to and including index `i`
      * 
      * @param i the last data index to force computation to
      */
    private[this] def computeToIndex(i: Int): Unit = {
      while (nextIndex <= i) {
        array(nextIndex) = iter.next()
        nextIndex += 1
      }
    }

    override def apply(i: Int): A = {
      require(i >= 0 && i < underlying.length)
      computeToIndex(i)
      array(i).asInstanceOf[A]
    }
    
    override def length: Int = underlying.length
    override def cached: IndexedSeqView[A] = this
  }
```


```scala
  // in object MapView...
  class Cached[K, +V](underlying: SomeMapOps[K, V]) extends MapView[K, V] {
    private[this] lazy val hashMap = underlying.to(immutable.HashMap)
    override def get(key: K): Option[V] = hashMap.get(key)
    override def iterator: Iterator[(K, V)] = Iterator.empty ++ hashMap
    override def cached: MapView[K, V] = this
    override def to[C1](factory: Factory[(K, V), C1]): C1 = hashMap.to(factory)
    override def toMap[K0, V0](implicit ev: (K, V) <:< (K0, V0)): immutable.Map[K0, V0] = 
      hashMap.toMap
  }
```

File: src/library/scala/collection/View.scala [hunk 1]
```diff
@@ -37,6 +35,12 @@ trait View[+A] extends Iterable[A] with IterableOps[A, View, View[A]] with Seria
 
   @deprecated("Views no longer know about their underlying collection type; .force always returns an IndexedSeq", "2.13.0")
   @`inline` def force: IndexedSeq[A] = toIndexedSeq
+
+  /**
+    * @return a view containing the elements of this view which only evaluates elements once,
+    *         and does so lazily.
+    */
+  def cached: View[A] = LazyList.from(this).view
```
Comment (smarter) [inline src/library/scala/collection/View.scala:43]: Why go through LazyList in particular ?

Comment (NthPortal) [reply src/library/scala/collection/View.scala]: it's lazy so it has good caching semantics - i.e. it doesn't require evaluating the entire view if you access a single element

Comment (NthPortal) [reply src/library/scala/collection/View.scala:43]: it's lazy so it has good caching semantics - i.e. it doesn't require evaluating the entire view if you access a single element.

also, I'm not sure what the alternative is

**Comment by NthPortal** [comment]
@joshlemer I'm actually struggling to implement a `cached` returning a `MapView` which is reasonably lazy. I'm not a fan of lazy evaluating everything at once

**Comment by NthPortal** [comment]
upon further thought based on the discussion here, I don't think just creating a `LazyList` of the elements and wrapping them is sufficient for `MapView` and `IndexedSeqView`, and I think it's a bit too late to be adding something that complex for 2.13

**Pull request closed without merging**